### ISSUE TITLE
build and remove images in parallel will get an error

### ISSUE THREAD

**Issue by JJfight1987** [containers/buildah#2925]
**Description**

I want to build and remove two images in parallel by 'buildah bud -t XXX' and 'buildah rmi xxx', But it always get an error when 'build rmi XXX' like this: 
error determining if the image "4ce292834e374b558737c3ef68e9df4818e6f09c77d8dcf2391890a3a7011832" is a parent:

How can I implement build -> push -> remove images in parallel in makefile (there is no push step in the test code)?

My test code only 2 files dockerfile and Makefile.
- dockerfile 
```
FROM scratch
```
- Makefile
```
.PHONY:  all
default: all

all:aa bb
	buildah images -a

aa:
	buildah bud -f Dockerfile -t aa ./
	buildah rmi aa

bb:
	buildah bud -f Dockerfile -t bb ./
	buildah rmi bb

clean:
	buildah rmi -af

show:
	buildah images -a
```
see attached [buildah-parallel.zip](https://github.com/containers/buildah/files/5848982/buildah-parallel.zip)

If use 'make -j 1', the two images can build and remove successfully;
If use 'make -j 2' always get an error.

**Steps to reproduce the issue:**
1.
2.
3.

**Describe the results you received:**

[jli11@ctu-hig1 /ctu-hig1_01/jli11/source/buildah]$make -j 2
buildah bud -f Dockerfile -t aa ./
buildah bud -f Dockerfile -t bb ./
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
WARN error reading allowed ID mappings: error reading subuid mappings for user "jli11" and subgid mappings for group "jli11": No subuid ranges found for user "jli11" in /etc/subuid
WARN Found no UID ranges set aside for user "jli11" in /etc/subuid.
WARN Found no GID ranges set aside for user "jli11" in /etc/subgid.
WARN error reading allowed ID mappings: error reading subuid mappings for user "jli11" and subgid mappings for group "jli11": No subuid ranges found for user "jli11" in /etc/subuid
WARN Found no UID ranges set aside for user "jli11" in /etc/subuid.
WARN Found no GID ranges set aside for user "jli11" in /etc/subgid.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 2: COMMIT aa
STEP 2: COMMIT bb
Getting image source signatures
Getting image source signatures
Copying blob 5f70bf18a086 done
Copying blob 5f70bf18a086 done
Copying config 8b2e8afb16 done
Writing manifest to image destination
Copying blob 5f70bf18a086 done
Copying config ef0a4a3b50 done
Writing manifest to image destination
Storing signatures
--> 8b2e8afb169
--> ef0a4a3b50a
8b2e8afb1695d13196d3307b1650eb7cc8a36aa6e167f17325c5bb60471c096f
ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8
buildah rmi aa
buildah rmi bb
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
WARN error reading allowed ID mappings: error reading subuid mappings for user "jli11" and subgid mappings for group "jli11": No subuid ranges found for user "jli11" in /etc/subuid
WARN Found no UID ranges set aside for user "jli11" in /etc/subuid.
WARN Found no GID ranges set aside for user "jli11" in /etc/subgid.
WARN error reading allowed ID mappings: error reading subuid mappings for user "jli11" and subgid mappings for group "jli11": No subuid ranges found for user "jli11" in /etc/subuid
WARN Found no UID ranges set aside for user "jli11" in /etc/subuid.
WARN Found no GID ranges set aside for user "jli11" in /etc/subgid.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
untagged: localhost/aa:latest
untagged: localhost/bb:latest
ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8
error removing image "8b2e8afb1695d13196d3307b1650eb7cc8a36aa6e167f17325c5bb60471c096f": unable to read configuration from image "ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8": unable to parse reference to image "ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8": error parsing named reference "ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8": invalid repository name (ef0a4a3b50aa8d56d9a2ca16806c04c32eb150808bdf58b73b2d77d3f8e71fb8), cannot specify 64-byte hexadecimal strings
ERRO exit status 125
Makefile:12: recipe for target 'bb' failed
make: *** [bb] Error 125
[jli11@ctu-hig1 /ctu-hig1_01/jli11/source/buildah]$buildah images -a
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
WARN error reading allowed ID mappings: error reading subuid mappings for user "jli11" and subgid mappings for group "jli11": No subuid ranges found for user "jli11" in /etc/subuid
WARN Found no UID ranges set aside for user "jli11" in /etc/subuid.
WARN Found no GID ranges set aside for user "jli11" in /etc/subgid.
ERRO[0000] The storage 'driver' option must be set in /etc/containers/storage.conf, guarantee proper operation.
REPOSITORY   TAG      IMAGE ID       CREATED          SIZE
<none>       <none>   8b2e8afb1695   14 seconds ago   2.11 KB

**Describe the results you expected:**

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah/unknown,now 1.19.2~1 amd64 [installed]
```

**Output of `buildah version`:**

```
Version:         1.19.2
Go Version:      go1.15.2
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        0.4.0
libcni Version:
image Version:   5.9.0
Git Commit:
Built:           Thu Jan  1 08:00:00 1970
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
(paste your output here)
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.4 LTS"
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

```

**Output of `uname -a`:**

```
Linux ctu-hig1 4.15.0-132-generic #136-Ubuntu SMP Tue Jan 12 14:58:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = ""

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

# Storage path for rootless users
#
# rootless_storage_path = "$HOME/.local/share/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to the UIDs/GIDs as they should appear outside of the container,
# and the length of the range of UIDs/GIDs.  Additional mapped sets can be
# listed and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and then a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped in-container ID,
# until all of the entries have been used for maps.
#
# remap-user = "containers"
# remap-group = "containers"

# Root-auto-userns-user is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid and /etc/subgid file.  These ranges will be partioned
# to containers configured to create automatically a user namespace.  Containers
# configured to automatically create a user namespace can still overlap with containers
# having an explicit mapping set.
# This setting is ignored when running as rootless.
# root-auto-userns-user = "storage"
#
# Auto-userns-min-size is the minimum size for a user namespace created automatically.
# auto-userns-min-size=1024
#
# Auto-userns-max-size is the minimum size for a user namespace created automatically.
# auto-userns-max-size=65536

[storage.options.overlay]
# ignore_chown_errors can be set to allow a non privileged user running with
# a single UID within a user namespace to run containers. The user can pull
# and use any image even those with multiple uids.  Note multiple UIDs will be
# squashed down to the default uid in the container.  These images will have no
# separation between the users in the container. Only supported for the overlay
# and vfs drivers.
#ignore_chown_errors = false

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev"

# Set to skip a PRIVATE bind mount on the storage home directory.
# skip_mount_home = "false"

# Size is used to set a maximum size of the container image.
# size = ""

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# Size is used to set a maximum size of the container image.
# size = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

```

**Comment by vrothberg** [comment]
Thanks for reaching out.

While I think users should not delete data during build, there may be space for improvement in c/storage.  I think it's possible to extend c/storage to prevent other processes from _deleting_ data.  Reading data, and adding new layers may continue working.

@nalind @giuseppe @rhatdan PTAL

**Comment by rhatdan** [comment]
This works for be on Fedora.
buildah -v
buildah version 1.18.0 (image-spec 1.0.1-dev, runtime-spec 1.0.2-dev)

**Comment by rhatdan** [comment]
```
make -j 2
buildah bud -f Dockerfile -t aa ./
buildah bud -f Dockerfile -t bb ./
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 2: COMMIT aa
STEP 2: COMMIT bb
Getting image source signatures
Copying blob 5f70bf18a086 done  
Getting image source signatures
Copying config c143990888 done  
Writing manifest to image destination
Storing signatures
Copying blob 5f70bf18a086 [--------------------------------------] 0.0b / 0.0b
Copying config 7739e22411 done  
Writing manifest to image destination
Storing signatures
--> c143990888e
c143990888e361ae3516610d609e6b4da5fbb5c5204b23f714076008e7fe93ff
--> 7739e224110
buildah rmi aa
7739e224110607eca650823740f82f7b20c4d5e31024e536437c709095c9fc15
buildah rmi bb
untagged: localhost/aa:latest
untagged: localhost/bb:latest
c143990888e361ae3516610d609e6b4da5fbb5c5204b23f714076008e7fe93ff
7739e224110607eca650823740f82f7b20c4d5e31024e536437c709095c9fc15
buildah images -a
REPOSITORY   TAG   IMAGE ID   CREATED   SIZE
```

**Comment by rhatdan** [comment]
Ok, this seems to be new in buildah 1.19.2
Older buildah this did not happen.

@nalind Ideas?

**Comment by JJfight1987** [comment]
> ```
> make -j 2
> buildah bud -f Dockerfile -t aa ./
> buildah bud -f Dockerfile -t bb ./
> STEP 1: FROM scratch
> STEP 1: FROM scratch
> STEP 2: COMMIT aa
> STEP 2: COMMIT bb
> Getting image source signatures
> Copying blob 5f70bf18a086 done  
> Getting image source signatures
> Copying config c143990888 done  
> Writing manifest to image destination
> Storing signatures
> Copying blob 5f70bf18a086 [--------------------------------------] 0.0b / 0.0b
> Copying config 7739e22411 done  
> Writing manifest to image destination
> Storing signatures
> --> c143990888e
> c143990888e361ae3516610d609e6b4da5fbb5c5204b23f714076008e7fe93ff
> --> 7739e224110
> buildah rmi aa
> 7739e224110607eca650823740f82f7b20c4d5e31024e536437c709095c9fc15
> buildah rmi bb
> untagged: localhost/aa:latest
> untagged: localhost/bb:latest
> c143990888e361ae3516610d609e6b4da5fbb5c5204b23f714076008e7fe93ff
> 7739e224110607eca650823740f82f7b20c4d5e31024e536437c709095c9fc15
> buildah images -a
> REPOSITORY   TAG   IMAGE ID   CREATED   SIZE
> ```

The reproduce may depends on machine performance, here is my machine cpu info:
```
[jli11@ctu-hig1 /ctu-hig1_01/jli11/source/buildah]$lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6240R CPU @ 2.40GHz
Stepping:            7
CPU MHz:             1000.029
BogoMIPS:            4800.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
```
I tested the test code on a 4-core machine it also hard to reproduce, but if increase the image number, it becomes more easy to reproduce.
Modify Makefile as below and use 'make -j 96(as big as possible)'
```
.PHONY:  all
default: all

all:aa bb cc dd ee ff
	buildah images -a

aa:
	buildah bud -f Dockerfile -t aa ./
	buildah rmi aa

bb:
	buildah bud -f Dockerfile -t bb ./
	buildah rmi bb

cc:
	buildah bud -f Dockerfile -t cc ./
	buildah rmi cc

dd:
	buildah bud -f Dockerfile -t dd ./
	buildah rmi dd
ee:
	buildah bud -f Dockerfile -t ee ./
	buildah rmi ee

ff:
	buildah bud -f Dockerfile -t ff ./
	buildah rmi ff

clean:
	buildah rmi -af

show:
	buildah images -a

```

**Comment by vrothberg** [comment]
I should have better parsed the error message. I _think_ that the shortname changes are at hand here.  I'll take a look.

**Comment by vrothberg** [comment]
> I should have better parsed the error message. I _think_ that the shortname changes are at hand here. I'll take a look.

No short names :)

**Comment by vrothberg** [comment]
Note that I can also reproduce with older Buildahs.

**Comment by vrothberg** [comment]
I see more errors than just the parse error.  Others relate to trying to remove an already removed image.

This could be a time sink and as we're planning to consolidate the image-handling code across Buildah and Podman, we may defer to that?

I tried reproducing the error with podman and did not manage.  A temporary workaround could be using `podman rmi` until we fixed buildah.

**Comment by JJfight1987** [comment]
I tried podman 2.2.1 also fails.
Makefile:
```
.PHONY:  all
default: all

all:aa bb cc dd ee ff
	podman images -a

aa:
	podman build -f Dockerfile -t aa ./
	podman rmi aa

bb:
	podman build -f Dockerfile -t bb ./
	podman rmi bb

cc:
	podman build -f Dockerfile -t cc ./
	podman rmi cc

dd:
	podman build -f Dockerfile -t dd ./
	podman rmi dd
ee:
	podman build -f Dockerfile -t ee ./
	podman rmi ee

ff:
	podman build -f Dockerfile -t ff ./
	podman rmi ff

clean:
	podman rmi -af

show:
	podman images -a
```
output log:
```
li11@jli11:~/workspace/parallel$ make -j 32
podman build -f Dockerfile -t aa ./
podman build -f Dockerfile -t bb ./
podman build -f Dockerfile -t cc ./
podman build -f Dockerfile -t dd ./
podman build -f Dockerfile -t ee ./
podman build -f Dockerfile -t ff ./
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 1: FROM scratch
STEP 2: COMMIT dd
STEP 2: COMMIT aa
STEP 2: COMMIT cc
STEP 2: COMMIT ff
STEP 2: COMMIT bb
STEP 2: COMMIT ee
--> e42ad519fbf
--> cf23225ca63
--> 2e6ad5de72d
--> 04fa4ce4929
e42ad519fbf9824c2087640e0cd066a0b9e0e9b1aa4f3c36ef98d8e259f36fe6
cf23225ca634bc3216e332d804dd50db22553fede05f3f205c9f94fc3e31e37a
2e6ad5de72dcacf0fe44c986d3e872e7822aa8c381ec96361a0aef09f16ca514
04fa4ce49296894d04103920d39e03103b8b3d1e17bbcf73c538bb3dcc0fdd6a
--> b9498eb7b93
podman rmi dd
podman rmi ff
--> bcbcaa5354e
podman rmi cc
podman rmi aa
b9498eb7b93e1747536813813d32eee27d1b117fa594fefb67a944bf83ee43ad
bcbcaa5354e08272674591ea3e7c402328228961477a1565ceeeb95f396f3989
podman rmi ee
podman rmi bb
Untagged: localhost/ff:latest
Deleted: e42ad519fbf9824c2087640e0cd066a0b9e0e9b1aa4f3c36ef98d8e259f36fe6
Untagged: localhost/dd:latest
Deleted: cf23225ca634bc3216e332d804dd50db22553fede05f3f205c9f94fc3e31e37a
Error: 1 error occurred:
	* error reading image "04fa4ce49296894d04103920d39e03103b8b3d1e17bbcf73c538bb3dcc0fdd6a" as image source: error reading image "04fa4ce49296894d04103920d39e03103b8b3d1e17bbcf73c538bb3dcc0fdd6a": error locating image with ID "04fa4ce49296894d04103920d39e03103b8b3d1e17bbcf73c538bb3dcc0fdd6a": image not known
Untagged: localhost/cc:latest
Deleted: 04fa4ce49296894d04103920d39e03103b8b3d1e17bbcf73c538bb3dcc0fdd6a
make: *** [Makefile:13: bb] Error 125
make: *** Waiting for unfinished jobs....
Untagged: localhost/aa:latest
Deleted: 2e6ad5de72dcacf0fe44c986d3e872e7822aa8c381ec96361a0aef09f16ca514
Untagged: localhost/ee:latest
Deleted: b9498eb7b93e1747536813813d32eee27d1b117fa594fefb67a944bf83ee43ad
```

**Comment by vrothberg** [comment]
Thanks @JJfight1987, that's super helpful.

In both cases, the `image removal` code must be improved.

**Comment by rhatdan** [comment]
I am assigning to @vrothberg

**Comment by vrothberg** [comment]
When reading the email I thought "Hey, but I just got that fixed" ... well in Podman :-)  Looking forward to consolidate the image code of Podman and Buildah.

That being said, please retry with the latest version of Podman.

**Comment by rhatdan** [comment]
Since buildah and podman share the same image handling now, I am going to assume this is fixed in the current release.
